### ISSUE TITLE
ValueError: Input contains NaN, infinity or a value too large for dtype('float64') after random exploration

### ISSUE THREAD

**Issue by bstemper** [scikit-optimize/scikit-optimize#594]
**Software-Versions:**
_OS_: Mac OSX 10.13.2
_Skopt_: 0.4

**How to get the error** My objective function _train_bayes_ is being minimised like such:

> res_gp = gp_minimize(train_bayes, space, n_calls=20, verbose=True, random_state=random_seed, n_jobs=-1)

Calling the objective function runs without error: All initial random explorations of the search space run without error. Only after the last iteration, when Bayesian updating takes place, the following error is raised: 
> ValueError: Input contains NaN, infinity or a value too large for dtype('float64')

It definitely can't be an issue with my objective function because

- changing the seed of my random objective function does not help
- setting n_random_starts=x leads to the same problem after the x initial explorations

**Example:**

> Iteration No: 1 started. Evaluating function at random point.
> Neural network does not learn.
> Iteration No: 1 ended. Evaluation done at random point.
> Time taken: 86.5439
> Function value obtained: 1.0000
> Current minimum: 1.0000
> Iteration No: 2 started. Evaluating function at random point.
> Neural network fully trained.
> Iteration No: 2 ended. Evaluation done at random point.
> Time taken: 67.2817
> Function value obtained: 0.0063
> Current minimum: 0.0063
> Iteration No: 3 started. Evaluating function at random point.
> Neural network does not learn.
> Iteration No: 3 ended. Evaluation done at random point.
> Time taken: 578.9058
> Function value obtained: 1.0000
> Current minimum: 0.0063
> Iteration No: 4 started. Evaluating function at random point.
> Neural network fully trained.
> Iteration No: 4 ended. Evaluation done at random point.
> Time taken: 19.8216
> Function value obtained: 0.0051
> Current minimum: 0.0051
> Iteration No: 5 started. Evaluating function at random point.
> Neural network does not learn.
> Iteration No: 5 ended. Evaluation done at random point.
> Time taken: 40.6203
> Function value obtained: 1.0000
> Current minimum: 0.0051
> Iteration No: 6 started. Evaluating function at random point.
> Neural network does not learn.
> Iteration No: 6 ended. Evaluation done at random point.
> Time taken: 555.5275
> Function value obtained: 1.0000
> Current minimum: 0.0051
> Iteration No: 7 started. Evaluating function at random point.
> Neural network does not learn.
> Iteration No: 7 ended. Evaluation done at random point.
> Time taken: 101.0601
> Function value obtained: 1.0000
> Current minimum: 0.0051
> Iteration No: 8 started. Evaluating function at random point.
> Neural network fully trained.
> Iteration No: 8 ended. Evaluation done at random point.
> Time taken: 31.3000
> Function value obtained: 0.0046
> Current minimum: 0.0046
> Iteration No: 9 started. Evaluating function at random point.
> Neural network does not learn.
> Iteration No: 9 ended. Evaluation done at random point.
> Time taken: 9.7913
> Function value obtained: 1.0000
> Current minimum: 0.0046
> Iteration No: 10 started. Evaluating function at random point.
> ---------------------------------------------------------------------------
> ValueError                                Traceback (most recent call last)
> <ipython-input-12-fda378d2b615> in <module>()
> ----> 1 res_gp = gp_minimize(train_bayes, space, n_calls=20, verbose=True, random_state=random_seed, n_jobs=-1)
> 
> ~/anaconda3/lib/python3.6/site-packages/skopt/optimizer/gp.py in gp_minimize(func, dimensions, base_estimator, n_calls, n_random_starts, acq_func, acq_optimizer, x0, y0, random_state, verbose, callback, n_points, n_restarts_optimizer, xi, kappa, noise, n_jobs)
>     217         n_restarts_optimizer=n_restarts_optimizer,
>     218         x0=x0, y0=y0, random_state=random_state, verbose=verbose,
> --> 219         callback=callback, n_jobs=n_jobs)
> 
> ~/anaconda3/lib/python3.6/site-packages/skopt/optimizer/base.py in base_minimize(func, dimensions, base_estimator, n_calls, n_random_starts, acq_func, acq_optimizer, x0, y0, random_state, verbose, callback, n_points, n_restarts_optimizer, xi, kappa, n_jobs)
>     249         fit_model = n < n_calls - 1
>     250         next_y = func(next_x)
> --> 251         result = optimizer.tell(next_x, next_y, fit=fit_model)
>     252         result.specs = specs
>     253 
> 
> ~/anaconda3/lib/python3.6/site-packages/skopt/optimizer/optimizer.py in tell(self, x, y, fit)
>     448             with warnings.catch_warnings():
>     449                 warnings.simplefilter("ignore")
> --> 450                 est.fit(self.space.transform(self.Xi), self.yi)
>     451 
>     452             if hasattr(self, "next_xs_") and self.acq_func == "gp_hedge":
> 
> ~/anaconda3/lib/python3.6/site-packages/skopt/learning/gaussian_process/gpr.py in fit(self, X, y)
>     193                 noise_level=self.noise, noise_level_bounds="fixed"
>     194             )
> --> 195         super(GaussianProcessRegressor, self).fit(X, y)
>     196 
>     197         self.noise_ = None
> 
> ~/anaconda3/lib/python3.6/site-packages/sklearn/gaussian_process/gpr.py in fit(self, X, y)
>     179         self._rng = check_random_state(self.random_state)
>     180 
> --> 181         X, y = check_X_y(X, y, multi_output=True, y_numeric=True)
>     182 
>     183         # Normalize target value
> 
> ~/anaconda3/lib/python3.6/site-packages/sklearn/utils/validation.py in check_X_y(X, y, accept_sparse, dtype, order, copy, force_all_finite, ensure_2d, allow_nd, multi_output, ensure_min_samples, ensure_min_features, y_numeric, warn_on_dtype, estimator)
>     571     X = check_array(X, accept_sparse, dtype, order, copy, force_all_finite,
>     572                     ensure_2d, allow_nd, ensure_min_samples,
> --> 573                     ensure_min_features, warn_on_dtype, estimator)
>     574     if multi_output:
>     575         y = check_array(y, 'csr', force_all_finite=True, ensure_2d=False,
> 
> ~/anaconda3/lib/python3.6/site-packages/sklearn/utils/validation.py in check_array(array, accept_sparse, dtype, order, copy, force_all_finite, ensure_2d, allow_nd, ensure_min_samples, ensure_min_features, warn_on_dtype, estimator)
>     451                              % (array.ndim, estimator_name))
>     452         if force_all_finite:
> --> 453             _assert_all_finite(array)
>     454 
>     455     shape_repr = _shape_repr(array.shape)
> 
> ~/anaconda3/lib/python3.6/site-packages/sklearn/utils/validation.py in _assert_all_finite(X)
>      42             and not np.isfinite(X).all()):
>      43         raise ValueError("Input contains NaN, infinity"
> ---> 44                          " or a value too large for %r." % X.dtype)
>      45 
>      46 
> 
> ValueError: Input contains NaN, infinity or a value too large for dtype('float64').
>

**Comment by betatim** [comment]
Interesting. From the traceback you posted it looks like it is the coordinates of the points that contain a Nan or Inf. Could you also post your search space configuration? Then we can try and reproduce it as I don't have an idea off the top of my head what is going wrong.

**Comment by bstemper** [comment]
Okay, so I was able to trace it down to the search space that I used:

> space = [(1,12), (1,12), (-7.0, -1), (3,11), (1,1)]

The last item _(1,1)_ was the issue. I wanted to keep the last parameter of my objective function fixed to 1, so that's why I used _(1,1)_ . This does work well when drawing from the space but the Bayesian updating somehow needs an interval with more than one element in it.

**Comment by MechCoder** [comment]
In that case why not fix the value to 1 in the objective function itself?

**Comment by betatim** [comment]
For me (on `master`) `Space(((1, 1),))` leads to an exception:
```python
~/git/scikit-optimize/skopt/space/space.py in __init__(self, low, high, transform, name)
    312         if high <= low:
    313             raise ValueError("the lower bound {} has to be less than the"
--> 314                              " upper bound {}".format(low, high))
    315         self.low = low
    316         self.high = high

ValueError: the lower bound 1 has to be less than the upper bound 1
```
to prevent people from accidentally creating a dimension like this. Would be good to see how you managed to create your space. Potentially somewhere there is a way around this check (which I'd consider a bug).

**Comment by bstemper** [comment]
@betatim From what you wrote, it looks like you explicitly created an instance of a 'space' class and then used that to call the optimizer (I haven't checked the code, so that could be wrong?). Anyway, what I did was only:
> space = [(1,12), (1,12), (-7.0, -1), (3,11), (1,1)]
so space is just a list of tuples. And then I called
> res_gp = gp_minimize(train_bayes, space, n_calls=20, verbose=True, random_state=random_seed)
If you tell me what you want me to look for, I can help maybe.

**Comment by betatim** [comment]
When you pass in a list of dimensions like this they should fairly quickly be converted into a `Space` instance. So the thing to investigate/trace is where things flow between entering `gp_minimize` and when the `Space` instance is created. At some point an instance of `Space` has to be created and should raise the same exception ... but from what you posted it does not appear like that is the case :-/

**Comment by iaroslav-ai** [comment]
Thank you @bstemper for letting us know about this issue!
This issue has been resolved with 0.5 release. For further details, see #617 .
